src/nodes.coffee

```diff
@@ -1812,14 +1815,7 @@ exports.Obj = class Obj extends Base
 
     # If this object is the left-hand side of an assignment, all its children
     # are too.
-    if @lhs
-      for prop in props when prop instanceof Assign
-        {value} = prop
-        unwrappedVal = value.unwrapAll()
-        if unwrappedVal instanceof Arr or unwrappedVal instanceof Obj
-          unwrappedVal.lhs = yes
-        else if unwrappedVal instanceof Assign
-          unwrappedVal.nestedLhs = yes
+    @propagateLhs()
```

Comment:
The biggest thing introduced here is to make the process of nodes (eg `Obj`/`Arr`/`Assign`) knowing whether they're on the left-hand-side of an assignment (ie setting `.lhs`/`.nestedLhs`) more shared/normalized/complete

Previously the recursive "LHS-self-awareness propagation" was being done in `compileNode()` (and was missing some cases eg splats). But we always need to know whether we're left-hand-side for AST generation as well (since different `type`s are used eg `ObjectExpression` vs `ObjectPattern`)

So `propagateLhs()` methods are used here to normalize this and allow the LHS-propagation to occur elsewhere besides just `compileNode()`

Reply:
It seems a little odd that we need a method to _set_ the left-handedness on this node’s children. I would think that the nodes would all have an `isOnLeftHandSide` method or similar, that first looks at itself for an `isLHS` property and then walks up the tree looking for such a property on each parent until we hit the first non-`Obj`/`Arr`/`Assign` node. But maybe I’m not fully understanding what’s going on here.

Reply:
@GeoffreyBooth I can picture that. But the existing way seems to be to go "top-down" and have parents responsible for telling their children that they're on the left-hand-side. This is just converting it from a more "ad-hoc" approach mixed into the traversal of children during `compileNode()` into a more clean, comprehensive method-based approach

And I'd guess that this way is a bit more compact (and definitely more efficient to the extent that matters)

Also our nodes don't have a reference to their parents (so no way to traverse up the tree) no?

Reply:
I guess you’re right. I’m looking at the traversal methods and they’re top-down like `traverseChildren`. So I guess this is the way it’s done.